test/junit/scala/collection/FactoriesTest.scala

```diff
@@ -105,6 +110,177 @@ class FactoriesTest {
+        s"SortedIterableFactory ($factory)'s method `from` should return the same reference when passed $input",
+        input,
+        factory.from(input))
+    }
+  def mapFactoryFromIterableOnceReturnsSameReference[CC[X, Y] <: Map[X, Y] , K, V](
+   factories: MapFactory[CC]*
+    )(
+     inputs: CC[K, V]*
+   ): Unit =
+    for {
+      factory <- factories
+      input <- inputs
+    } {
+      assertSame(
+        s"MapFactory ($factory)'s method `from` should return the same reference when passed $input",
+        input,
+        factory.from(input))
+    }
+  def sortedMapFactoryFromIterableOnceReturnsSameReference[CC[X, Y] <: Map[X, Y] , K, V](
+    factories: SortedMapFactory[CC]*
+  )(
+    inputs: CC[K, V]*
+  )(implicit ordering: Ordering[K]): Unit =
+    for {
+      factory <- factories
+      input <- inputs
+    } {
+      assertSame(
+        s"SortedMapFactory ($factory)'s method `from` should return the same reference when passed $input",
+        input,
+        factory.from(input))
+    }
+
+
+  @Test
+  def testFactoriesReuseCollectionsWhenPossible(): Unit = {
+
+    factoryFromIterableOnceReturnsSameReference(Iterable, im.Iterable)(
+      SortedSet(1, 2, 3),
+      im.SortedSet(1, 2, 3),
+      im.TreeSet(1, 2, 3),
+      im.TreeMap(1 -> 1, 2 -> 1, 3 -> 1).keySet,
+      new Enumeration {
+        val x, y, z = Value
+      }.values,
+      im.BitSet(1, 2, 3),
+      im.Map(1 -> 2),
+      im.HashMap(1 -> 2),
+      im.IntMap(1 -> 2),
+      im.LongMap(1L -> 2)
+    )
+
+    factoryFromIterableOnceReturnsSameReference(Iterator)(
+      Iterator(1),
+      List(1, 2, 3).iterator
+    )
+
+    factoryFromIterableOnceReturnsSameReference(LinearSeq, im.LinearSeq, Seq, im.Seq, Iterable, im.Iterable)(
+      im.LinearSeq(),
+      List(1, 2, 3),
+      Stream(1, 2, 3),
+      LazyList(1, 2, 3),
+      im.Queue(1, 2, 3)
+    )
+
+    factoryFromIterableOnceReturnsSameReference(IndexedSeq, im.IndexedSeq, Seq, im.Seq, Iterable, im.Iterable)(
+      Vector(),
+      Vector(1, 2, 3),
+      im.ArraySeq(1, 2, 3),
+      1 to 5,
+      "hello"
+    )
+
+    factoryFromIterableOnceReturnsSameReference(Set, im.Set, Iterable, im.Iterable)(
+      im.Set(1),
+      im.HashSet("a", "b", "c"),
+      im.ListSet('c', 'd'),
+      im.Map("a" -> 1, "b" -> 1, "c" -> 1).keySet,
+      im.HashMap("a" -> 1, "b" -> 1, "c" -> 1).keySet,
+    )
+
+    sortedFactoryFromIterableOnceReturnsSameReference(SortedSet, im.SortedSet)(
+      SortedSet(1, 2, 3),
+      im.SortedSet(1, 2, 3),
+      im.TreeSet(1, 2, 3),
+      im.TreeMap(1 -> 1, 2 -> 1, 3 -> 1).keySet,
+      im.BitSet(1, 2, 3)
+    )
+
+    /* TODO: When ord0 == ord1 returns true, uncomment this
+    object enum extends Enumeration {
+      val x, y, z = Value
+    }
+
+    val enumValues = enum.values
+
+    val ord0 = enumValues.ordering
+    val ord1 = implicitly[Ordering[enum.Value]]
+
+    sortedFactoryFromIterableOnceReturnsSameReference(SortedSet, im.SortedSet)(enumValues)
+    */
```

Comment:
what's up with this commented out block?

Reply:
Currently, the ordering that is implicitly available on enumeration.Value, is a low priority Ordering that uses implicit conversions from `Comparable[T]`. However, the ordering used by the SortedSet `enum.values` uses a different ordering (`Enumeration#ValueOrdering`). So then, when you call `SortedSet.from(enum.values)` it picks up a different ordering than the one that that sortedSet uses. So it rebuilds the SortedSet with the low-priority ordering.

Reply:
So, that will fail the test because the two orderings are new referentially equal (nor `==` equal)

Reply:
@NthPortal see https://github.com/scala/scala/pull/7591

Reply:
No longer commented out, since #7591 is merged.